tensorflow/contrib/opt/python/training/model_average_optimizer.py

```diff
@@ -0,0 +1,693 @@
+    # the accumulator to be global step. This list contains list of the
+    # following format: (accumulator, device).
+    self._accumulator_list = []
+    # ModelAverageHook should be called before ReplicasHook.
+    self._ma_run_hook = False
+    # name: string. Name of the global variables and related operation on ps.
+    self._name = 'modelAverage'
+
+    self.generate_local_and_global_variables()
+
+  def generate_local_and_global_variables(self):
+    # Change all variables to local variables.
+    for v in variables.global_variables():
+      if v.op.name.find(self._name) != -1:
+        raise AssertionError('%s: cannot use \'%s\' as a substr of any name for'
+                             ' ops or variables when calling an '
+                             'ModelAverageOptimizer.' % (v.op.name, self._name))
+      ops.add_to_collection(ops.GraphKeys.LOCAL_VARIABLES, v)
+    # Clear global_variables list.
+    ops.get_default_graph().clear_collection(ops.GraphKeys.GLOBAL_VARIABLES)
+    # Generate new global variables dependent on trainable variables.
+    for i, v in enumerate(variables.trainable_variables()):
+      if v.op.name.find(self._name) != -1:
+        raise AssertionError('%s: cannot use \'%s\' as a substr of a name for '
+                             'any ops or variables when calling an '
+                             'ModelAverageOptimizer.' % (v.op.name, self._name))
+      # v_g is the global-variable version of each user-defined trainable
+      # variable.  They are supposed to be placed on PS device. v_g is used to
+      # store averaged model parameters.
+      v_g = variable_scope.get_variable(name='%s_g%d' % (self._name, i),
+                                  initializer=v.initialized_value(),
+                                  trainable=False,
+                                  collections = [ops.GraphKeys.GLOBAL_VARIABLES,
+                                                 'global_model'])
+      v_block_grad = variable_scope.get_variable(
+                  name='%s_block_grad%d' % (self._name, i), shape=v.get_shape(),
+                  initializer=init_ops.constant_initializer(0.0, dtype=v.dtype),
+                  trainable=False,
+                  collections = [ops.GraphKeys.GLOBAL_VARIABLES, 'block_grads'])
+      v_nesterov = variable_scope.get_variable(
+                  name='%s_nesterov%d' % (self._name, i),
+                  initializer=v_g.initialized_value(),
+                  trainable=False,
+                  collections = [ops.GraphKeys.GLOBAL_VARIABLES, 'nesterov'])
+    self._super_global_step = variables.Variable(0,
+                  name="%s_global_step" % self._name, trainable=False)
+    self._num_of_global_variables = len(variables.global_variables())
+    self._num_of_trainable_variables = len(variables.trainable_variables())
+
+  def apply_model_average(self,
+                          lvars_and_gvars,
+                          global_vars,
+                          v_vars,
```

Comment:
v_vars seems to be the same as block_grads? Why does it have a different name here?